zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/common/Connector.java

```diff
@@ -0,0 +1,261 @@
+                        break;
+                }
+            } else {
+                if (path != null && path.startsWith(zNodePath)) {
+                    // Something has changed on the node, let's find out.
+                    try {
+                        zooKeeper.exists(path, true, this, null);
+                    } catch (Exception e) {
+                        throw new EndpointGroupException("Failed to process a ZooKeeperProxy watch event", e);
+                    }
+                }
+
+            }
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    break;
+                case NONODE:
+                    break;
+                case SESSIONEXPIRED:
+                    // Ignore this and let the zNode Watcher process it first.
+                case NOAUTH:
+                    // It's possible that this happens during runtime. We ignore this and wait for the ACL
+                    // configuration returns to normal. If it happens when the ZooKeeper client is initially
+                    // constructed, the constructor will throw an exception.
+                    return;
+                default:
+                    // Retry on recoverable errors. Fatal errors go to the process() method above.
+                    try {
+                        zooKeeper.exists(path, true, this, null);
+                    } catch (Exception ex) {
+                        throw new ZooKeeperException("Failed to process ZooKeeperProxy callback event", ex);
+                    }
+                    return;
+            }
+
+            if (!activeClose) {
+                nodeChange(zooKeeper, path);
+                //enqueue an end flag to force the main thread to wait until this callback finished  before exit
+                if (stateQueue != null) {
+                    enqueueState(KeeperState.Disconnected);
+                }
+            }
+
+        }
+
+        /**
+         * Enqueue the state.
+         * @param state ZooKeeperProxy state
+         */
+        private void enqueueState(KeeperState state) {
+            if (stateQueue == null) {
+                return;
+            }
+            try {
+                stateQueue.put(state);
+            } catch (InterruptedException e) {
+                throw new EndpointGroupException("Failed to enqueue the state.", e);
+            }
+        }
+    }
+
+    @VisibleForTesting
+    public BlockingQueue<KeeperState> stateQueue() {
+        return stateQueue;
+    }
+
+    /**
+     * Open state recording.
+     */
+    @VisibleForTesting
+    public void enableStateRecording() {
+        if (stateQueue == null) {
+            stateQueue = new ArrayBlockingQueue<>(10);
+        }
+    }
+
+    @VisibleForTesting
+    public ZooKeeper underlyingClient() {
```

Comment:
If this is visible for testing, could it be package-local so that it's not part of public API?